streams/src/main/java/org/apache/kafka/streams/kstream/Produced.java

```diff
@@ -0,0 +1,163 @@
+     * @return  A new {@link Produced} instance configured with keySerde
+     * @see KStream#through(String, Produced)
+     * @see KStream#to(String, Produced)
+     */
+    public static <K, V> Produced<K, V> keySerde(final Serde<K> keySerde) {
+        return new Produced<>(keySerde, null, null);
+    }
+
+    /**
+     * Create a Produced instance with provided valueSerde.
+     * @param valueSerde    Serde to use for serializing the key
+     * @param <K>           key type
+     * @param <V>           value type
+     * @return  A new {@link Produced} instance configured with valueSerde
+     * @see KStream#through(String, Produced)
+     * @see KStream#to(String, Produced)
+     */
+    public static <K, V> Produced<K, V> valueSerde(final Serde<V> valueSerde) {
+        return new Produced<>(null, valueSerde, null);
+    }
+
+    /**
+     * Create a Produced instance with provided partitioner.
+     * @param partitioner   the function used to determine how records are distributed among partitions of the topic,
+     *                      if not specified and the key serde provides a {@link WindowedSerializer} for the key
+     *                      {@link WindowedStreamPartitioner} will be used&mdash;otherwise {@link DefaultPartitioner} will be used
+     * @param <K>           key type
+     * @param <V>           value type
+     * @return  A new {@link Produced} instance configured with partitioner
+     * @see KStream#through(String, Produced)
+     * @see KStream#to(String, Produced)
+     */
+    public static <K, V> Produced<K, V> streamPartitioner(final StreamPartitioner<? super K, ? super V> partitioner) {
+        return new Produced<>(null, null, partitioner);
+    }
+
+    /**
+     * Produce records using the provided partitioner.
+     * @param partitioner   the function used to determine how records are distributed among partitions of the topic,
+     *                      if not specified and the key serde provides a {@link WindowedSerializer} for the key
+     *                      {@link WindowedStreamPartitioner} will be used&mdash;otherwise {@link DefaultPartitioner} wil be used
+     * @return this
+     */
+    public Produced<K, V> withStreamPartitioner(final StreamPartitioner<? super K, ? super V> partitioner) {
+        this.partitioner = partitioner;
+        return this;
+    }
+
+    /**
+     * Produce records using the provided valueSerde.
+     * @param valueSerde    Serde to use for serializing the value
+     * @return this
+     */
+    public Produced<K, V> withValueSerde(final Serde<V> valueSerde) {
+        this.valueSerde = valueSerde;
+        return this;
+    }
+
+    /**
+     * Produce records using the provided keySerde.
+     * @param keySerde    Serde to use for serializing the key
+     * @return this
+     */
+    public Produced<K, V> withKeySerde(final Serde<K> keySerde) {
+        this.keySerde = keySerde;
+        return this;
+    }
+
+    public Serde<K> keySerde() {
```

Comment:
I'm wondering whether we should expose these three functions (and the same for all other options classes of KIP-182). My gut feeling is not to, and probably we can do sth. similar to KIP-120 to abstract these functions for internal usage only. WDYT? @dguy @bbejeck @mjsax .

Reply:
Here is a more concrete idea: we can have an `o.a.k.streams.kstream.internals.StreamOptions` class, and let all the public `o.a.k.streams.kstream` classes have one `StreamOptions` as a `package private` field. This class just maintain a setter / getter of all the possible specs like serde / reset policy / timestamp extractor / partitioner etc. And then inside `o.a.k.streams.kstream.internals.XXImpl` we can access this object and get the required fields.

Reply:
i'll look into it. I agree with you about exposing the getters. I'd rather not

Reply:
Actually that won't work as the classes are in the wrong packages, i.e, `o.a.k.streams.kstream` vs `o.a.k.streams.kstream.internals` so the classes in `kstream.internals` wouldn't have access to the `package private` fields of the classes in `kstream`

Reply:
One option is that we have subclasses of the option classes in internal. We make the fields in the option classes protected and they would also have a protected copy ctor, i.e, 
`protected Produced(final Produced produced)`
We would then have in `internals`
```
class ProducedInternal<K, V> extends Produced<K, V> {
     public ProducedInternal(final Produced<K, V> produced) {
         super(produced)
     }

    public Serde<K> keySerde() {
       return super.keySerde; 
   }
  // etc
}
```
So, when we get called with `Produced` we do `new ProducedInternal(produced)` and we can then access the fields and remove the getters from the option classes - WDYT?

Reply:
I like the subclass idea. Sounds like a clean design to me.

Reply:
Agreed, probably a pattern we could carry forward for limiting the exposure of Streams internals.

Reply:
Sounds good to me, thanks @dguy

Reply:
The copy ctor needs to be public as it is in a different package. I said `protected`, but in the example i had `public` - doh. I still think this is better than having the getters?